walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
+        if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
+            potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));      
+        }
+      } else {
+        // Time slots have been found.
+        // Remove the last timerange from list (time is prevEvent.end to endOfDay) to check conflict with currEvent.
+        potentialTime.remove(potentialTime.size()-1);
+        
+        // Check to see if prev and curr events OVERLAP.
+        // Add timerange from the end of the later timerange.
+        if(prevEvent.getWhen().overlaps(currEvent.getWhen())){
+          int laterEnd = currEvent.getWhen().end();
+          if(prevEvent.getWhen().end() > laterEnd) {
+            laterEnd = prevEvent.getWhen().end();
+          }
+          potentialTime.add(TimeRange.fromStartEnd(laterEnd, TimeRange.END_OF_DAY,true));
+        } else {
+          // check GAP (...) in-between curr and prev.
+          // |--prev--|...|--curr--|
+          long gapDuration = currEvent.getWhen().start() - prevEvent.getWhen().end();
+          if(gapDuration >= request.getDuration()){
+            potentialTime.add(TimeRange.fromStartEnd(prevEvent.getWhen().end(), currEvent.getWhen().start(), false));
+          }
+          // Check there there is a gap between the end of the event and end of day large enough for request.
+          if(currEvent.getWhen().end() != TimeRange.END_OF_DAY){
+            if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
+              potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));  
+            } 
+          }
+        }
+      }
+    }
+    return potentialTime;
+  }
+  /**
+   * Finds all events that mandatory and optional attendees are booked for - removes duplicates.
+   * @param mandatory - events that mandatory attendees in the request have to attend.
+   * @param optional - events that optional attendees in the request have to attend.
+   * @return list of events.
+   */
+  public List<Event> getAllUniqueEvents(List<Event> mandatory, List<Event> optional) {
+
+    List<Event> allEvents = new ArrayList<>(mandatory);
+
+    for(Event opt : optional){
+      if(!allEvents.contains(opt)){
+        allEvents.add(opt);
+      }
+    }
+    return allEvents;
+  } 
+  /**
+   * Basic Bubble sorting of event list so they are in ascending order by start time.
+   * @param events - list of events to be compared by start time.
+   * @return sorted list of events.
+   */
+  public List<Event> sortEventsByStartTime(List<Event> events) {
```

Comment:
While it's nice to see you got some practice writing a bubble sort, `Collections.sort` should get the job done without having to implement your own sorting algorithm. It's usually a good idea to use a library sorting function because it's easy to introduce subtle errors when rolling your own.

Reply:
Done